File: src/lib/models/Transaction.ts
// this model should be desined to answer Who,What,When,Where,Why, and how
// Who performed the transaction
// Who recieved the transaction (if possible)
// Where did the money come from?
// When was the money spent
// When was the money recieve (if possible)
// Where was the money spent
// What was the money spent on
// How was the transaction paid?

// this model is to implement only functionallity required on the JS side
export class Transaction {
    // where did the money come from
    originatorId:string
    // who or what was the money sent to
    recieverId:string
    // who or what performed the transaction
    senderId:string  | null
    // where was the money sent from
    originatingLocation: string | null
    // where was the money sent to
    recievingLocation: string | null
    // where was the location of the person or thing who performed this transaction
    sendingLocation: string | null
    // what was the money spent on
    items: string | null
    // the total cost/gain of this transaction
    amount: number
    // what currency was the transaction performed in
    originatingCurrency:string
    // the currency of the reciever. used to allow conversion to/from external currencies
    recievingCurrency:string
    // what account the money was sent from
    originatorAccount:string
    // what account the money was sent to
    recieverAccount:string
    // used to organize transactions
    categories: string | null
}

File: entity/src/transaction.rs
use sea_orm::entity::prelude::*;
#[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
#[sea_orm(table_name = "transaction")]
pub struct Model {
    #[sea_orm(primary_key)]
    /// identifies this transaction
    id: Uuid,
    /// where did the money come from
    originator_id: Uuid,
    /// who or what was the money sent to
    reciever_id: Uuid,
    /// who or what performed the transaction
    sender_id: Option<Uuid>,
    /// where was the money sent from
    originating_location: Option<String>,
    /// where was the money sent to
    recieving_location: Option<String>,
    /// where was the location of the person or thing who performed this transaction
    sending_location: Option<String>,
    /// what was the money spent on
    items: Option<String>,
    /// the total cost/gain of this transaction
    amount: i64,
    // what currency was the transaction performed in
    originating_currency: String,
    // the currency of the reciever. used to allow conversion to/from external currencies
    recieving_currency: String,
    // what account the money was sent from
    originator_account: String,
    // what account the money was sent to
    reciever_account: String,
    // used to organize transactions
    categories: Option<String>,
    // when was the transaction created
    created_date: DateTimeWithTimeZone,
    // when was the transaction finalized
    finalized_date: Option<DateTimeWithTimeZone>,
    // whether the transaction is pending, returned, or complete
    status: String,
    // whether the transaction was handled with card,cash or ACH
    method: String,
}

impl ActiveModelBehavior for ActiveModel {}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::user::Entity")]
    User,
}

impl Related<super::user::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}
